clusterloader2/pkg/measurement/util/histogram.go

```diff
@@ -17,17 +17,70 @@ limitations under the License.
 package util
 
 import (
+	"math"
 	"reflect"
+	"sort"
+	"strconv"
 
 	"github.com/prometheus/common/model"
 )
 
+// Bucket of a histogram, copy from https://github.com/kubernetes/kubernetes/pull/85861/files#diff-8fc16044937fc03c39b585a657b20150R179-R209
+type bucket struct {
+	upperBound float64
+	count      float64
+}
+
+func bucketQuantile(q float64, buckets []bucket) float64 {
+	if q < 0 {
+		return math.Inf(-1)
+	}
+	if q > 1 {
+		return math.Inf(+1)
+	}
+
+	if len(buckets) < 2 {
+		return math.NaN()
+	}
+
+	rank := q * buckets[len(buckets)-1].count
+	b := sort.Search(len(buckets)-1, func(i int) bool { return buckets[i].count >= rank })
+
+	if b == 0 {
+		return buckets[0].upperBound * (rank / buckets[0].count)
+	}
+
+	// linear approximation of b-th bucket
+	brank := rank - buckets[b-1].count
+	bSize := buckets[b].upperBound - buckets[b-1].upperBound
+	bCount := buckets[b].count - buckets[b-1].count
+
+	return buckets[b-1].upperBound + bSize*(brank/bCount)
+}
+
 // Histogram is a structure that represents distribution of data.
 type Histogram struct {
 	Labels  map[string]string `json:"labels"`
 	Buckets map[string]int    `json:"buckets"`
 }
 
+// Quantile calculates the quantile 'q' based on the given buckets of Histogram.
+func (h *Histogram) Quantile(q float64) (float64, error) {
+	var b []bucket
+	for k, v := range h.Buckets {
```

Comment:
I see. `h.Buckets` is a map so the individual buckets are processed randomly. You need to make sure the list of  buckets that goes into `bucketQuantile` is sorted by bucket's upper bound. That's why I created `Validate` function to make sure the buckets are properly sorted.

Reply:
yeah, the buckets parameter passed to the bucketQuantile function is not sorted by bucket's upperbound, but i think we should sort list of buckts in bucketQuantile to ensure the correctness of result.

Reply:
Implementation of `bucketQuantile` you copy-pasted is private. In my case all the buckets are always properly sorted. So it's safe to call it as it is. In your case where the source histogram buckets are represented as map, it's your implementation detail to make sure the buckets are properly sorted.

Reply:
@ingvagabund if we create a ```NewHistogram``` function to create a [```Histogram```](https://github.com/kubernetes/kubernetes/pull/87923/files#diff-be7d56a59338e31acbaf032075732116R188) instance, then this PR can import your implementation instead of copying it.

Reply:
+1 for it, just finishing the unit tests

Reply:
This is the best I can offer right now: https://github.com/kubernetes/kubernetes/pull/87923/files#diff-be7d56a59338e31acbaf032075732116R234

`Histogram` data type under `k8s.io/component-base/metrics/testutil` is a wrapper around prometheus's Histogram. It needs to be a wrapper so I can clear all histogram samples and reset it to its initial state. Thus, `GetHistogramFromBuckets` will accept a list of buckets (which is generic enough to be used for various histogram source formats). However, computing average will give nonsensical values since we don't have the samples to get overall sample sum.